packages/endo/README.md

```diff
@@ -0,0 +1,283 @@
+modules (ESM) that do not depend upon any powerful built-in modules or
+facilities of global scope at the command line.
+
+Future versions may support:
+
+* CommonJS modules,
+* Passing powerful and powerless built-in modules and globals into specific
+  compartments based on application policies,
+* Realm isolation,
+* Running vetted shims in a realm before [lockdown].
+* Threading endowments,
+* Threading built-in modules,
+* Membranes or tamper-proofing between packages,
+* Extensions for module and program translation scoped to each compartment,
+* Running Endo applications on the web without a build step.
+* Generalized support for workers under Node or on the web.
+* An `importmap` superset.
+
+  [compartment]: https://github.com/Agoric/SES-shim/blob/master/packages/ses/README.md#compartment
+  [lockdown]: https://github.com/Agoric/SES-shim/blob/master/packages/ses/README.md#lockdown
+
+
+## Usage
+
+Endo runs your application in a single frozen realm, endowing only the main
+compartment with the given powerful global properties and built-in modules.
+
+The `importPath` function runs a compartmentalized application off the file
+system.
+The `endowments` are properties to add to the `globalThis` in the global scope
+of the application's main package compartment.
+The `modules` are built-in modules to grant the application's main package
+compartment.
+
+```js
+import "ses";
+import fs from "fs";
+import { importPath } from "endo";
+
+lockdown();
+
+// ...
+
+const modules = { fs };
+const endowments = { console };
+
+const { namespace } = await importPath(
+  fs.promises.readFile,
+  modulePath,
+  endowments,
+  modules
+);
+```
+
+> TODO
+>
+> A future version will allow application authors to distribute
+> their global endowments and granted built-in modules to third-party
+> packages within the application, à la [LavaMoat].
+
+The `importPath` function uses `loadPath`.
+Using `loadPath` directly allows for deferred execution or multiple runs with
+different endowments or modules.
+
+Use `writeArchive` to capture an application in an archival format.
+
+```js
+import fs from "fs";
+import { writeArchive } from "endo";
+
+await writeArchive(
+  fs.promises.writeFile,
+  fs.promises.readFile,
+  'app.agar', // the archive to write
+  'app.js' // the application to capture
+);
+```
+
+The `writeArchive` function uses `makeArchive`.
+Using `makeArchive` directly gives you the archive bytes.
+
+Use `importArchive` to run an application from an archive.
+Note the similarity to `importPath`.
+
+```js
+import "ses";
+import fs from "fs";
+import { importPath } from "endo";
+
+lockdown();
+
+// ...
+
+const modules = { fs };
+const endowments = { console };
+
+const { namespace } = await importArchive(
+  fs.promises.readFile,
+  archivePath,
+  endowments,
+  modules
+);
+```
+
+The `importArchive` function composes `loadArchive` and `parseArchive`.
+Use `loadArchive` to defer execution or run multiple times with varying
```

Comment:
I'm guessing we'll eventually need something to load an archive from disk and report something about it's properties. I'm thinking about the `modules` and `endowments`: is there anything in the archive which expresses the application's desires? Or is the loader supposed to just somehow know what it wants?

It's not a trivial question, and maybe not something to figure out now. The caller of `importArchive` is obviously in charge: they decide what powers to grant to the newly-imported code. But there's a certain amount of conversation taking place, basically between the author of the archive and the party doing the importing. The author says "my code needs X to do its job", and the importer looks at that (and perhaps also signatures or recommendations from delegated auditors, and maybe some interaction with the human whose data is at stake), and decides whether it's ok granting those requests or not. Or substituting in something else (app wants geolocation data, human says "ok but low-res", importer gives zipcode but nothing better). When I look at the archive with my old (dusty) Jetpack hat on, I think of it as a container for the author's intentions, both their code and the "headers" that talk about that code, including how it wants to be wired up.

For now, I imagine that the human must read the code and find the `import` statements that refer to things which aren't in the archive, make a decision about them, and then instruct the importing program to add those names to `modules` and/or `endowments`. Which is probably plenty for this early stage.

Reply:
This is a great point. The compartment map should eventually be able to express the endowments and built-in modules it expects and we should have a place in the API to surface that, like `StaticModuleRecord` is both bindable-then-executable and describes a module’s static dependencies. I can change the archive from being a closure to an object with a distinguished method. We might not just surface the names of the needed things, but also eventually their types.

Reply:
sounds great for this stage